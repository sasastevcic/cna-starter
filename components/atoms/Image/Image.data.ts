import { ImageProps as NextImageProps } from 'next/image';

export type AspectRatio = [number, number];

export type AspectRatioProps = Pick<NextImageProps, 'layout' | 'width' | 'height'>;

export interface ImageProps extends NextImageProps {
	aspectRatio?: AspectRatio;
}

export interface StyledImageProps {
	$aspectRatio?: AspectRatio;
	$hasFit?: boolean;
}
