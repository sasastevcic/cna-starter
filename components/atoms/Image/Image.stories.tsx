import { ComponentStory, ComponentMeta } from '@storybook/react';
import ImageComponent from './index';

export default {
	title: 'UI Elements/Image',
	component: ImageComponent,
} as ComponentMeta<typeof ImageComponent>;

const Template: ComponentStory<typeof ImageComponent> = (args) => <ImageComponent {...args} />;

export const Cover = Template.bind({});
Cover.args = {
	src: 'https://via.placeholder.com/150',
	alt: 'Placeholder',
	objectFit: 'cover',
	layout: 'fill',
};

export const Contain = Template.bind({});
Contain.args = {
	src: 'https://via.placeholder.com/150',
	alt: 'Placeholder',
	objectFit: 'contain',
	width: 240,
	height: 160,
};

export const AspectRatio = Template.bind({});
AspectRatio.args = {
	src: 'https://via.placeholder.com/150',
	alt: 'Placeholder',
	aspectRatio: [16, 9],
	layout: 'fill',
	objectFit: 'cover',
};
