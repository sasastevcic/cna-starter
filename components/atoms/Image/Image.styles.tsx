import { size } from 'polished';
import styled from 'styled-components';
import { aspectRatio } from '../../../styles/helpers/aspectRatio';
import type { StyledImageProps } from './Image.data';

export const StyledImage = styled.div<StyledImageProps>`
	${size('auto', '100%')};

	${({ $aspectRatio }) => $aspectRatio && aspectRatio($aspectRatio)};
`;
