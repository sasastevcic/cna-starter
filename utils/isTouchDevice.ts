import { isBrowser } from './isBrowser';

export const isTouchDevice = (): boolean => {
	if (isBrowser) {
		return 'ontouchstart' in window || navigator.maxTouchPoints > 0;
	}

	return false;
};
