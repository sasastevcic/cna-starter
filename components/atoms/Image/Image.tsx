import { ComponentType, ReactElement } from 'react';
import NextImage from 'next/image';
import { ImageProps } from 'next/image';
import { AspectRatioImageProps } from './Image.data';
import { StyledAspectRatio } from './Image.styles';

const withAspectRatio =
	(Component: ComponentType<ImageProps>) =>
	({ aspectRatio, ...props }: AspectRatioImageProps) => {
		if (aspectRatio) {
			return (
				<StyledAspectRatio $aspectRatio={aspectRatio}>
					<Component layout="fill" objectFit="cover" {...props} />
				</StyledAspectRatio>
			);
		}

		return <Component {...props} />;
	};

const Image = (props: AspectRatioImageProps): ReactElement => (
	<NextImage data-testid="NextImage" quality={100} {...props} />
);

export default withAspectRatio(Image);
