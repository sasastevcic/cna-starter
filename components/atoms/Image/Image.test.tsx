import renderer from 'react-test-renderer';
import { renderWithTheme } from '../../../utils/testsWithTheme';
import Image from './index';

const imageProps = {
	alt: 'background',
	src: 'https://picsum.photos/200',
};

describe('Image', () => {
	describe('Render', () => {
		it('should render', () => {
			const { getByTestId } = renderWithTheme(<Image src={imageProps.src} alt={imageProps.alt} />);

			const element = getByTestId('Image');

			expect(element).toBeInTheDocument();
		});

		it('should render correctly', () => {
			const tree = renderer.create(<Image src={imageProps.src} alt={imageProps.alt} />).toJSON();

			expect(tree).toMatchSnapshot();
		});
	});

	describe('Attributes', () => {
		it('should have correct alt text', () => {
			const { getByTestId } = renderWithTheme(<Image src={imageProps.src} alt={imageProps.alt} />);

			const element = getByTestId('ImageElement');

			expect(element).toHaveAttribute('alt', imageProps.alt);
		});

		it('should have correct source', () => {
			const { getByTestId } = renderWithTheme(<Image src={imageProps.src} alt={imageProps.alt} />);

			const element = getByTestId('ImageElement');

			expect(element).toHaveAttribute('src', imageProps.src);
		});
	});
});
