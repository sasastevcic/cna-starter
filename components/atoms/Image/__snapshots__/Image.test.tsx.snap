// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image Render should render correctly 1`] = `
<span
  style={
    {
      "background": "none",
      "border": 0,
      "bottom": 0,
      "boxSizing": "border-box",
      "display": "block",
      "height": "initial",
      "left": 0,
      "margin": 0,
      "opacity": 1,
      "overflow": "hidden",
      "padding": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
      "width": "initial",
    }
  }
>
  <img
    alt="background"
    data-nimg="fill"
    data-testid="NextImage"
    decoding="async"
    onError={[Function]}
    onLoad={[Function]}
    src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
    style={
      {
        "border": "none",
        "bottom": 0,
        "boxSizing": "border-box",
        "display": "block",
        "height": 0,
        "left": 0,
        "margin": "auto",
        "maxHeight": "100%",
        "maxWidth": "100%",
        "minHeight": "100%",
        "minWidth": "100%",
        "objectFit": undefined,
        "objectPosition": undefined,
        "padding": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "width": 0,
      }
    }
  />
  <noscript>
    <img
      alt="background"
      data-nimg="fill"
      data-testid="NextImage"
      decoding="async"
      loading="lazy"
      sizes="100vw"
      src="/_next/image?url=https%3A%2F%2Fpicsum.photos%2F200&w=3840&q=100"
      srcSet="/_next/image?url=https%3A%2F%2Fpicsum.photos%2F200&w=640&q=100 640w, /_next/image?url=https%3A%2F%2Fpicsum.photos%2F200&w=750&q=100 750w, /_next/image?url=https%3A%2F%2Fpicsum.photos%2F200&w=828&q=100 828w, /_next/image?url=https%3A%2F%2Fpicsum.photos%2F200&w=1080&q=100 1080w, /_next/image?url=https%3A%2F%2Fpicsum.photos%2F200&w=1200&q=100 1200w, /_next/image?url=https%3A%2F%2Fpicsum.photos%2F200&w=1920&q=100 1920w, /_next/image?url=https%3A%2F%2Fpicsum.photos%2F200&w=2048&q=100 2048w, /_next/image?url=https%3A%2F%2Fpicsum.photos%2F200&w=3840&q=100 3840w"
      style={
        {
          "border": "none",
          "bottom": 0,
          "boxSizing": "border-box",
          "display": "block",
          "height": 0,
          "left": 0,
          "margin": "auto",
          "maxHeight": "100%",
          "maxWidth": "100%",
          "minHeight": "100%",
          "minWidth": "100%",
          "objectFit": undefined,
          "objectPosition": undefined,
          "padding": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": 0,
        }
      }
    />
  </noscript>
</span>
`;
